fix(masthead): display L0 nav with L1 when scrolling up

emyarod opened the pull request:
### Related Ticket(s)

#6391

related https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/97

### Description

This PR updates the Dotcom Shell Composite web component and React Masthead behavior to show the L0 nav when scrolling up if an L1 nav is also present

In the Dotcom Shell Composite web component, this is done by changing the Table of Contents component's `stickyOffset` **property** on scroll if an L1 is present. The `stickyOffset` **attribute** is not changed though, so one point of discussion for the reviewers would be whether or not we also want the attribute to reflect the property. If we do, then I believe we'll need to store the original attribute value the shell composite for calculations

In the React Masthead component, this is done by setting the sticky masthead classes based on scroll direction if an L1 is present. The actual offset values are on the application side so the storybook content component needs to know whether or not there is an L1 to pass the correct offset value to `<TableOfContents />`

https://user-images.githubusercontent.com/8265238/140102486-23fbb3b2-5ec5-4634-b158-ba9d89ad0a94.mov

### Changelog

**New**

- `withL1` prop on `<Content />`

**Changed**

- Dotcom Shell Composite masthead and table of contents position styles
- React Masthead sticky styles logic

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

oliviaflory commented:
Hey @emyarod overall this looks really good! Adding some observations that may or may not be in scope, we can break out to other issues if needed.

React
1. On Desktop, using the TOC to scroll back up, the headline is cut off Desktop. It doesn't do that in the prod storybook, but the L0 also doesn't come into view in prod today. 
<img width="1389" alt="Screen Shot 2021-11-04 at 3 38 23 PM" src="https://user-images.githubusercontent.com/45692622/140416600-587aa114-9b6c-4465-9043-4b69600162c7.png">


![Desktop React scroll upNov-04-2021 15-39-27](https://user-images.githubusercontent.com/45692622/140416573-d8f6a219-28fc-43c5-8d5b-900ecfc4e193.gif)


Pretty sure this isn't in scope because it's in prod as well. Happening in both React and Web components

On medium or small breakpoint, if I interact with the TOC while it's over the Feature card, I am taken to the feature card's link. Same thing happens over the Call out with media, the video plays rather than opening the native drop down with the TOC.
![TOC-callout--Nov-04-2021 15-56-00](https://user-images.githubusercontent.com/45692622/140419309-2c4f76e9-f4c5-400c-b79b-badb16f82ff0.gif)

emyarod commented:
> React
>
> 1. On Desktop, using the TOC to scroll back up, the headline is cut off Desktop. It doesn't do that in the prod storybook, but the L0 also doesn't come into view in prod today.

@oliviaflory this should be resolved now!

> Pretty sure this isn't in scope because it's in prod as well. Happening in both React and Web components
> 
> On medium or small breakpoint, if I interact with the TOC while it's over the Feature card, I am taken to the feature card's link. Same thing happens over the Call out with media, the video plays rather than opening the native drop down with the TOC.

I fixed this already in https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/7548! it hasn't been released yet but it should be working correctly in Canary

oliviaflory commented:
@emyarod thank you for the update!

1. React: The cut off headline is definitely fixed on Desktop breakpoints 👍 
~I do see a bit of variation in cut off vs visible on the medium and small breakpoints where the TOC changes to the select. Showing the variations here with two scroll down and one scroll up, maybe the variation depends on how each band/section is set up within the Dotcom shell story?~   Scratch all that, the TOC should be the only sticky element in the medium and small breakpoints when the TOC is present on a page. Lmk if this is not in scope, we could address in another issue. 

2. 
> I fixed this already in #7548! it hasn't been released yet but it should be working correctly in Canary

Yay!

3. Additionally, now when using the TOC the 'active' item and the headline are disconnected. If I click on the TOC item "Lobortis elementum nibh tellus" I scroll to the correct content block, but the active TOC item is the section above "Tincidunt ornare massa". I hope that makes sense, including a GIF as well.

Not sure if that is in scope for this PR? 

![TOC-item--Nov-09-2021 15-45-53](https://user-images.githubusercontent.com/45692622/141001912-05334682-3fd9-416e-9558-da278a0552d0.gif)


4. Web components: I see a bit of variation in cut off vs visible on the medium and small breakpoints where the TOC changes to the select. And a bit of disconnect of which item/section is active.
Maybe the variation depends on how each band/section is set up within the Dotcom shell story and the variation in spacing between section? If this isn't in scope LMK, seems like the shell might need a bit of cleaning up.
<img width="953" alt="Screen Shot 2021-11-09 at 4 10 18 PM" src="https://user-images.githubusercontent.com/45692622/141005471-2d2e1e36-600c-42f1-905a-8e1afc0c3689.png">
<img width="956" alt="Screen Shot 2021-11-09 at 4 10 33 PM" src="https://user-images.githubusercontent.com/45692622/141005474-d80720e4-2468-47e6-8b78-050067cc8678.png">

emyarod commented:
@oliviaflory I believe I addressed all of the points in your last review! but if 1 and 4 still persist then they should probably be tracked separately since it's slightly out of the scope of this issue

IgnacioBecerra commented on packages/styles/scss/components/masthead/_masthead.scss:
```diff
@@ -213,25 +213,26 @@ $search-transition-timing: 95ms;
+  .#{$prefix}--masthead--sticky__l1
+    + .#{$prefix}--dotcom-shell
+    .#{$prefix}--tableofcontents__sidebar {
+    top: 98px;
```
Might be better using the `rem()` function in the stylesheet to keep things consistent across styles?

emyarod replied:
@IgnacioBecerra should I convert the existing styles there to `rem` as well?
